fix: Guard against attribute errors in TesseractOcrModel __del__

bbrowning opened the pull request:
This moves the initialization of the `reader` and `script_readers` attributes to before we attempt to import tesserocr, so that when later accessing these attributes in the garbage collection method `__del__` the attributes exist.

This requires changing the typing of the `script_readers` dict value to `Any` because we cannot yet reference its actual strong type, since it's a tesserocr value.

This prevents throwing an exception during garbage collection for cases where the TesseractOcrModel instance didn't properly initialize, like when it throws an `ImportError` during its initializer.


**Issue resolved by this Pull Request:**
Resolves #1493

booxter commented:
This will do. How about just unconditionally setting it to None in `__init__`, which was the case before some refactoring happened?

dolfim-ibm commented:
I agree that setting those variables to `None` in the `__init__` is better. I think `self.reader` would work easily, the `self.script_readers` might raise typing issues, since the tesserocr imports are later.

bbrowning commented:
I hesitated to update where those attributes are initialized initially since they were moved recently to their current location. But, I agree it's simpler to initialize these earlier and get rid of the `hasattr` checks, so I updated and rebased this to do that.

The one thing I'll call out is this did require me to loosen up the type check of the `script_readers` dict values, since the previous typed value was a tesserocr reference that we haven't imported yet. The new value it types to is just `Any`, which is loose but avoids the chicken-and-egg issue of not being able to initialize this to the proper strong type until after tesserocr is imported.

dolfim-ibm commented:
Thanks @bbrowning !

booxter commented:
There's a way to defer annotations resolution, see: https://github.com/docling-project/docling/pull/1500
